### PR TITLE
feat(ui): Unbind Keybinds

### PR DESCRIPTION
**Pull request by Koranir**
**Feature:** This PR implements the feature request detailed and discussed in issue #8332

## Feature Details
Pressing x or Delete over a keybind in controls will disable it.

## UI Screenshots

https://user-images.githubusercontent.com/101683475/229463836-b33f63b4-95f1-4794-81fa-1f2f4fda365a.mp4

### CODE REVIEW

**Comment by Terin** [comment]
Two keys sharing a keybind, including being unbound, causes a crash upon entering the preferences panel.

**Comment by Koranir** [comment]
> Two keys sharing a keybind, including being unbound, causes a crash upon entering the preferences panel.

I'm not getting that crash, are there any other specifics?

**Comment by Terin** [comment]
You can clear them out while the game is running, but closing and reopening the panel, or relaunching the game will cause the crash. Running in debug mode is even less stable for me, where clicking a bind while immediately throw an exception.
image

Perhaps it could just be an issue with my Windows build environment, I'll try on Linux to see if it's the same there.

**Comment by Koranir** [comment]
Confirmed this issue.

File: source/PreferencesPanel.cpp [hunk 1]
```diff
@@ -408,7 +418,7 @@ void PreferencesPanel::DrawControls()
 			// Mark conflicts.
 			bool isConflicted = command.HasConflict();
 			bool isEditing = (index == editing);
-			if(isConflicted || isEditing)
+			if((isConflicted || isEditing))
```
Comment (Hurleveur) [inline source/PreferencesPanel.cpp:421]:
```suggestion
			if(isConflicted || isEditing)
```

File: source/PreferencesPanel.cpp [hunk 2]
```diff
@@ -140,6 +140,8 @@ bool PreferencesPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &comma
 		++currentSettingsPage;
 	else if((key == 'r' || key == SDLK_PAGEDOWN) && currentSettingsPage > 0)
 		--currentSettingsPage;
+	else if(key == 'x' || key == SDLK_DELETE)
+		Command::SetKey(zones[latest].Value(), 0);
```
Comment (Hurleveur) [inline source/PreferencesPanel.cpp:144]: This doesnt prevent assigning a key to 0, does it?

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Not sure why we need to make a new behavior with latest but I dont know this part of the code well

File: source/PreferencesPanel.cpp (hunk 2 above)
Comment (Koranir) [reply source/PreferencesPanel.cpp:144]: IIRC It's the SDL_Keycode 0, which is undefined and not on a keyboard, so cannot be assigned otherwise.

Comment (Hurleveur) [reply source/PreferencesPanel.cpp:144]: We can still give a value to 'x', right? How does that work?

Comment (Koranir) [reply source/PreferencesPanel.cpp:144]: https://wiki.libsdl.org/SDL2/SDLKeycodeLookup

Here at 0x00

Comment (Hurleveur) [reply source/PreferencesPanel.cpp:144]: Oh nvm I get it this works when you've unselected the keybind which is why we need to store the latest one.

**Comment by Koranir** [comment]
> Not sure why we need to make a new behavior with latest but I dont know this part of the code well

It's a workaround to support unbinding for both mouse and keyboard navigation, based on what was latest. If there are any better solutions, I'm open for discussion.

**Review by Hurleveur — APPROVED** [review]
looks good to me code wise

**Comment by samrocketman** [comment]
Love the idea.  Minor feedback.  ES tends to reuse shortcuts depending on context which is convenient for gamepads and steam deck.  For example,

* D deploys fighters in space but is Done in menus.
* B boards ships in space but Buys in outfitter and shipyard.

X is a fairly unique keybind in ES.  For game controllers that have limited buttons it would be nice to reuse shortcuts from existing gameplay keybinds.  For example, on gamepad X is typically used to select an enemy ship (R shortcut).  So R could be the keybind to "Remove shortcut"

**Comment by Hurleveur** [comment]
We could also just have delete as a key for this

**Comment by samrocketman** [comment]
> We could also just have delete as a key for this (and not consider it an exception like is done rn but instead considering setting a key to del as removing it) Idk whats the gamepad equivalent of del tho

This would still pose the same problem for gamepads.  Limited buttons means there's an advantage for shortcut reuse piggyback on other commonly used shortcuts.  Changing their meaning depending on context (for example in this case the context would be R in the shortcut menu would remove shortcuts but R means something else in several other game contexts).

**Comment by quyykk** [comment]
I would suggest adding something like `(None)` to that it is clear that the action is unassigned.

File: source/Command.cpp [hunk 3]
```diff
@@ -209,9 +209,24 @@ const string &Command::Description() const
...
 	auto it = keyName.find(*this);
-	return (it == keyName.end() ? empty : it->second);
+
+	if(it == keyName.end())
+		return false;
+
+	if(it->second == "")
+		return false;
+
+	return true;
```
Comment (quyykk) [inline source/Command.cpp:229]: This can be a single return.

File: source/Command.cpp [hunk 4]
```diff
@@ -209,9 +209,24 @@ const string &Command::Description() const
...
+// Check if the key has no binding.
+bool Command::HasBinding() const
+{
 	auto it = keyName.find(*this);
-	return (it == keyName.end() ? empty : it->second);
+
+	if(it == keyName.end())
+		return false;
+
+	if(it->second == "")
```
Comment (quyykk) [inline source/Command.cpp:226]: Use `.empty()`

File: source/PreferencesPanel.cpp [hunk 5]
```diff
@@ -140,6 +140,13 @@ bool PreferencesPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &comma
 		++currentSettingsPage;
 	else if((key == 'r' || key == SDLK_PAGEDOWN) && currentSettingsPage > 0)
 		--currentSettingsPage;
+	else if((key == 'x' || key == SDLK_DELETE) && (page == 'c'))
+	{
+		if(zones[latest].Value().KeyName() != Command::MENU.KeyName())
+		{
```
Comment (quyykk) [inline source/PreferencesPanel.cpp:146]: drop these braces

File: data/interfaces.txt [hunk 6]
```diff
@@ -417,6 +418,12 @@ interface "controls"
 		dimensions 90 30
 		color medium
 		hover bright
+	label "Unbind highlighted with 'x'"
```
Comment (quyykk) [inline data/interfaces.txt:421]:
```suggestion
	label "Unbind highlighted keybind with 'x'"
```

I didn't check if it fits

File: source/Command.cpp (hunk 4 above)
Comment (Koranir) [reply source/Command.cpp]: How?

File: source/Command.cpp [hunk 7]
```diff
@@ -212,9 +212,21 @@ const string &Command::Description() const
...
 const string &Command::KeyName() const
 {
-	static const string empty;
+	static const string empty = "(none)";
 	auto it = keyName.find(*this);
-	return (it == keyName.end() ? empty : it->second);
+
+	return (!HasBinding() ? empty : it->second);
+}
+
```
Comment (warp-core) [inline source/Command.cpp:220]:
```suggestion



```

File: source/Command.cpp [hunk 8]
```diff
@@ -212,9 +212,21 @@ const string &Command::Description() const
...
+// Check if the key has no binding.
+bool Command::HasBinding() const
+{
+	auto it = keyName.find(*this);
+
+	if(it == keyName.end() || it->second == "")
+		return false;
+
+	return true;
```
Comment (warp-core) [inline source/Command.cpp:229]: If we're going to combine the conditions into a single expression, just skip the if statements and return the result of the expression directly.
If we're using if statements for readability, then the conditions should be simpler expressions.

Pick one:

```suggestion
	if(it == keyName.end())
		return false;
	if(it->second.empty())
		return false;
	return true;
```

```suggestion
	return !(it == keyName.end() || it->second.empty());
```

File: data/_ui/interfaces.txt [hunk 9]
```diff
@@ -143,6 +143,7 @@ color "planet target pointer dominated" 1. 1. 1. 0.
 color "error back" .25 .1 .1 1.
 color "warning back" .21 .18 .08 1.
 color "warning conflict" .315 .27 .12 1.
+color "warning nocommand" 0.16 0.16 0.13 1.
```
Comment (warp-core) [inline data/_ui/interfaces.txt:146]:
```suggestion
color "warning no command" 0.16 0.16 0.13 1.
```

File: source/PreferencesPanel.cpp [hunk 10]
```diff
@@ -352,8 +357,17 @@ void PreferencesPanel::DrawControls()
 	const Color &medium = *GameData::Colors().Get("medium");
 	const Color &bright = *GameData::Colors().Get("bright");
 
-	// Check for conflicts.
+	// Colors for highlighting.
 	const Color &warning = *GameData::Colors().Get("warning conflict");
+	const Color &nocommand = *GameData::Colors().Get("warning nocommand");
```
Comment (warp-core) [inline source/PreferencesPanel.cpp:362]:
```suggestion
	const Color &noCommand = *GameData::Colors().Get("warning no command");
```

File: source/PreferencesPanel.cpp [hunk 11]
```diff
@@ -434,11 +448,12 @@ void PreferencesPanel::DrawControls()
...
 			// Mark conflicts.
 			bool isConflicted = command.HasConflict();
+			bool isEmpty = !command.HasBinding();
 			bool isEditing = (index == editing);
-			if(isConflicted || isEditing)
+			if(isConflicted || isEditing || isEmpty)
 			{
 				table.SetHighlight(56, 120);
-				table.DrawHighlight(isEditing ? dim : warning);
+				table.DrawHighlight(isEditing ? dim : isEmpty ? nocommand : warning);
```
Comment (warp-core) [inline source/PreferencesPanel.cpp:456]:
```suggestion
				table.DrawHighlight(isEditing ? dim : isEmpty ? noCommand : warning);
```

**Pull request merged**